### DISCUSSION TITLE
Plugin Ideas and Schema Contributions

### DISCUSSION THREAD

**Discussion by hand-dot** [pdfme/pdfme#288] [Ideas]
You can propose ideas for plugins that can be used in pdfme V3 and beyond while discussing them here.

If there is a need for certain features that should be supported, we can create a separate issue for them. However, this is a good place to have various exchanges of opinions.

If you've created a schema, please do report it here. Your plugin should be able to solve problems for the community and new users.

---

Document - Custom Schemas(Plugins) : https://pdfme.com/docs/custom-schemas

**Comment by dkastl** [comment]
Thank you for your work on additional schema support.

I'm interested in printing maps with pdfme, where maps would be embedded in the PDF together with other content.

**Reply by hand-dot** [reply]
That's an interesting idea. I'd be happy if you could go ahead and implement it, then report back.

**Reply by dkastl** [reply]
I was able to work on a `MapImage` plugin, following the plugin documentation page and the signature example. Thanks for all these examples, which were really helpful!

I published an initial version in my project fork: https://github.com/dkastl/pdfme/tree/ftr/map-schema

The plugin creates an interactive map (when set to editable) using OpenLayers. It has two simple fields to specify a point location, which will be drawn as a point feature on the map. In future versions I want to make this more flexible, but for now this was just a test.

![image](https://github.com/pdfme/pdfme/assets/227762/921db411-2b2a-4d8b-beb2-a39074fe6988)

There are a few things I'm struggling with:

(1) There is no `defaultValue` for maps as dataUrl, so I need to create one right after the map has been created for the first time. However, `onChange(data)` only works in "edit" mode, and after some try & error I could get the best result by writing to the `schema.data` property: https://github.com/dkastl/pdfme/blob/ftr/map-schema/playground/src/plugins/map/index.ts#L17-L19 Not sure if this is a good idea. Having a few events to listen to, or a function to update the `data` would be really helpful. There are still cases where the dataUrl of the map is not what is shown on the screen (probably an old version of the dataUrl).

(2) Some map layers require time to fetch map tiles for example. So I needed to delay creating the dataUrl. Otherwise the canvas will be empty. It would be very convenient, if update of the dataUrl could be initiated for example before print generation.

(3) It seems the UI completely rebuilds every time there is a change. So the map element also rebuilds completely. There are cases where I had multiple concurrent calls, and in combindation with waiting for tiles to load this can cause lots of headaches.

if someone finds such a plugin useful, let me know. If you have advice how to write things better or any other feedback, that's very appreciated.

My goal is to replace an implementation using https://mapfish.github.io/mapfish-print-doc/#/overview with PDFme. It's part of another Open Source project: https://gtt-project.org/

**Reply by hand-dot** [reply]
Hey @dkastl ,

First of all, Thank you for creating plugin!!
I'm happy to see such an interesting plugin. It was worth creating the mechanism of the plugin.

I will answer your questions.

Regarding 1,
This is a kind of hack, because the viewer only renders based on the input value(latitude and longitude).

Regarding 2,
I'm not sure if I fully understand your intention, but I don't think passing a base64 image generated in the UI as an input is a good method. My advice is basically the same as above, but you should generate images based on input value(latitude and longitude) and then render them in the PDF.

Regarding 3,
Could you please show me the specific steps where re-rendering occurs and a screen recording?

**Comment by hand-dot** [comment]
I have created a schema for the signature.
![signature-schema](https://github.com/pdfme/pdfme/assets/24843808/e10e91f2-7c7b-433e-bedf-dbe673883d20)

Playground: [playground.pdfme.com](https://playground.pdfme.com/)
Code: https://github.com/pdfme/pdfme-playground/blob/main/src/plugins/signature.ts

**Reply by peteward** [reply]
this is very cool!

**Reply by Stepykun** [reply]
Hello @hand-dot . I'm trying to put an empty signature boxes using this example, but I get an error:
pdfHelper.js:6 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'startsWith')
    at isPdf (pdfHelper.js:6:1)
    at pdf (image.js:15:1)
    at generate (generate.js:51:1)
 Any ideas on how to generate pdf with empty signature?

**Reply by hand-dot** [reply]
I think this was already fixed by this https://github.com/pdfme/pdfme/issues/498

**Reply by Stepykun** [reply]
@hand-dot Indeed. I updated to the latest version and the issue is gone.
Arigato!

**Comment by peteward** [comment]
In pdfme v2, we forked our version so that we could allow editors to mix dynamic and static content within a single text schema:
![Screenshot 2023-10-30 at 07 16 08](https://github.com/pdfme/pdfme/assets/7068515/39e9a221-3267-460b-8ed4-e173392439e7)

We will look to re-implement this within v3 as a custom schema. However, we will consider ways to improve both UX and functionality.

One idea I had was to allow editors to introduce multiple variables into a single text schema simply by surrounding them with `{` `}` when typing, e.g.:

```
Hi {name}, thanks for buying {product}, we hope you enjoy it.
```

* This allows the style, sizing and spacing of the text with multiple dynamic variables
* Also just static text schema with no variables at all.

Unfortunately, I think this is probably not possible right now as the relationship of input vars to schema is probably 1-to-1. 
I will investigate this further in the coming weeks and see what we can do.

**Reply by hand-dot** [reply]
How about accepting a JSON string as an input variable, and then parsing it into an object for internal expansion during rendering?

**Reply by peteward** [reply]
good idea! 
this would mean we couldn't share variables between other schemas, but I don't think this is a problem.

**Comment by hand-dot** [comment]
MEMO
- readOnlyText
- line
- rectangle
- circle

**Reply by hand-dot** [reply]
ref: https://github.com/pdfme/pdfme/issues/320

**Comment by dkastl** [comment]
Not sure if this makes sense, but a set of "common" types could ensure that input data is formatted correctly. 

With "common types" I mean for example:

- Numbers (which can be even distinguish between `integer` or `float` for example)
- Boolean ... maybe this could be rendered as checkboxes for example
- Date/time (timestamp), duration

Or others like:

- Hyperlink 
- Currency
- List (simple or key/value) or enums ... could be shown as dropdowns

Do you already have some sort of documentation how to correctly add plugins/schema?

**Reply by hand-dot** [reply]
> Do you already have some sort of documentation how to correctly add plugins/schema?

Yes, https://pdfme.com/docs/custom-schemas

**Reply by hand-dot** [reply]
ref: https://github.com/pdfme/pdfme/issues/319

**Comment by AdamD2L** [comment]
Hi ... I would need a checkbox schema :) ... thank you

**Reply by hand-dot** [reply]
Hi @AdamD2L,

Could you please clarify the specifications and create an issue for it? For instance, if a checkbox box is prepared on the PDF, there may be several options, such as wanting to draw only the check.

I would like you to suggest something that meets your needs and is easy for other users to use as well. Once the specifications are clear, it seems possible to implement them.

**Reply by AdamD2L** [reply]
Hi, thank you @hand-dot.
The user can insert a checkbox field
The user can choose to insert the box or just the tick symbol
The user can choose the tick symbol : tick, cross, ...
The user can change the size and position of the check box with mouse or by entering values
The user can choose the color of the tick symbol and the color of the box border and background
The user can change the alignment of the checkbox

It would be cool to have that ! Thank you !

**Comment by LeandroBanaszeski** [comment]
Hi... I had a need in the image schema, I made a change to meet it, an opacity option in the images, a practical example, like a watermark. It would be interesting to be in the Git Hub source code, so I don't need to control the source separately.

Change follows:
![Captura de tela 2023-11-10 103752](https://github.com/pdfme/pdfme/assets/45818324/47d8352d-2a89-47c8-9a54-0e421d734810)

![image](https://github.com/pdfme/pdfme/assets/45818324/df5573d9-4ca3-4398-95af-09872677b190)

![Captura de tela 2023-11-10 104010](https://github.com/pdfme/pdfme/assets/45818324/32e26c06-d437-4d97-9558-8b01bf344604)

![Captura de tela 2023-11-10 104056](https://github.com/pdfme/pdfme/assets/45818324/c53a0545-b92f-403c-a00b-615cab3d889d)

Result:

![image018](https://github.com/pdfme/pdfme/assets/45818324/2d83201a-50f6-4ca7-8a9e-2591f41005c9)

**Reply by hand-dot** [reply]
@LeandroBanaszeski 
Great, I think it's excellent. Could you create a PR? I would be happy to review it and merge it.

**Comment by jedimonkey** [comment]
I've started some dev on making a form text field plugin, it's like a cut down text plugin, but means the output is editable in the final pdf. Is the intent to make plugins stand alone, or move them into the main pdfme/plugins module?
I noticed that the barcode plugin is in the playground code, not plugins. There is a couple of things that would be handy if they are exported from the plugins code base, so I can reuse existing functionality and types if you new plugins to be external.

**Reply by hand-dot** [reply]
Hey @jedimonkey ,


We plan to integrate into the pdfme/schemas module only those things that are highly versatile and desired by many users, and only if developers contribute to pdfme.

No, the barcode plugin is not in the playground. It's located here: https://github.com/pdfme/pdfme/tree/main/packages/schemas/src/barcodes

> There is a couple of things that would be handy if they are exported from the plugins code base, so I can reuse existing functionality and types if you new plugins to be external.

What exactly are you referring to? Please let me know, and I can export them as needed.

**Reply by jedimonkey** [reply]
whoops! not barcode, I meant signature! 
All good - what you are saying makes sense. I'll work out what bits make sense to export and then can give you a list to discuss. :)

also - keep up the great work! This is a brilliant library, I've built pdf writers from scratch before (in php, many many years ago) and can appreciate how much of a pita they can be!

**Reply by hand-dot** [reply]
Thank you! I'll be waiting, so feel free to post again whenever you like.

**Reply by Mystificious** [reply]
@jedimonkey Hey Marcus, would you mind sharing your plugin-code for the editable fields in the final PDF?

**Comment by giovanni-orciuolo** [comment]
Would be really nice to have support for checkboxes and radioboxes! Going further, it would be even cooler to have support for conditional rendering of additional static data/form inputs based on the state of such checkboxes/radioboxes.

**Reply by hand-dot** [reply]
Hi @giovanni-orciuolo 
Since version 5.2.0, [expressions](https://pdfme.com/docs/expression) are now supported. Could you check it out?  
Checkboxes and radio buttons are planned for future development.
- https://github.com/pdfme/pdfme/issues/570
- https://github.com/pdfme/pdfme/issues/457

**Comment by vpn-dev** [comment]
Hey @hand-dot do you have any plugin coded which is like a row container having no. of columns like flex row and children will be reltive to that parent container

**Reply by hand-dot** [reply]
Hey @vpn-dev 
No, I don't have one.
What kind of thing are you trying to make?

**Reply by vpn-dev** [reply]
[Kazam_screencast_00005.webm](https://github.com/pdfme/pdfme/assets/37739139/a180fe71-a2db-486f-bd57-2d6bdb36ab01)

@hand-dot as you can see in above video I am adding one roq with 2 columns and in those columns separate fields like image and heading , as shown in video now column has some properties like padding or align-content which also affect children fields/widgets..

this kind of thing i am saying its like html builder kind of thing

**Comment by omorShahriar** [comment]
hi @hand-dot , you are doing some amazing work there. I am a freelancer and using pdfme in one of my ongoing projects. Your library is the nicest abstraction that meet our needs. We understand that pdfme is under ongoing development. here are some features we would love to have. Fillable checkboxes, options, calendar date picker. We are thinking of creating our own plugins for the time being. If anyone from the community have such solutions already built, please do share. That would be really helpful.

**Reply by hand-dot** [reply]
Hi @omorShahriar 

Latest pdfme, we can use Date / Time / DateTime schemas
https://pdfme.com/docs/supported-features#date--time--datetime

**Comment by kingsnow129** [comment]
Hi @hand-dot, Here are 3 requirements I'm not sure it's already accomplished.

1. Generate PDF with dynamic pictures
For example
There is a checklist, and each check item may have none or multiple pictures.
And the requirement is to export all the check items and pictures in one PDF.
2. Report header/footer which the whole PDF has one header at the beginning and one footer at the end. And Page header/footer which will shown for each page.
3. According to the data hide/show some fields, don't need to create different templates for various scenarios.

**Reply by hand-dot** [reply]
Hi @kingsnow129 

2. currently we can use [header and footer](https://pdfme.com/docs/headers-and-footers)
3. this will be achieved by [expression](https://pdfme.com/docs/expression)

**Comment by gmartini20** [comment]
I was looking for a way to create a field that would be editable after the PDF is exported similar to what @jedimonkey mentioned [here](https://github.com/pdfme/pdfme/discussions/288#discussioncomment-7593764). I wasn't able to find support to it on the lib or any plugin. Posting it here in case someone knows a plugin or a workaround to achieve it. Thanks

**Reply by hand-dot** [reply]
Hi @gmartini20 
Maybe related this issue: https://github.com/pdfme/pdfme/issues/67

**Comment by AyushBobale** [comment]
# Variable font/Bold, italics, bold-italic support
Hii @hand-dot, 
I am trying to achieve the functionality of "bold", "italic" and "bold-italic" at a field level.
In a previous attempt of mine in version 2, I had made a hacky way where I would change fonts to achieve this.
Along with that I had multiple font support i.e Arial, Roboto, Helvetica, etc. Along with their bold, italic, bold-italic versions.
Reason being pdf-lib not natively supporting any of this functionality.

My end goal is to add a schema for rich-text field.
So is there any natively supported functions to handle variable fonts in pdfme/pdf-lib, if not are you planning to add them any time soon.
**I am assuming that variable fonts support is quite difficult and  will not be added anytime soon**
If not then a basic functionality of adding in bold, italic and bold-italic variants of a font.

Current implementation in v5.2.10
```ts
const fontObjList = [
  {
    fallback: true,
    label: 'NotoSerifJP-Regular',
    url: '/fonts/NotoSerifJP-Regular.otf',
  },
  {
    fallback: false,
    label: 'NotoSansJP-Regular',
    url: '/fonts/NotoSansJP-Regular.otf',
  },
  {
    fallback: false,
    label: DEFAULT_FONT_NAME,
    data: getDefaultFont()[DEFAULT_FONT_NAME].data,
  },
];
```

Desired implementation [would you consider this implementation reasonable] 
``` ts
// variants would be optional
// if a variant is not found fallback to regular font
const fontObjList = [
  {
    fallback: true,
    label: 'NotoSerifJP',
    url: '/fonts/NotoSerifJP-Regular.otf'
    variants: {
        bold: '/fonts/NotoSerifJP-Bold.otf'
        italic: '/fonts/NotoSerifJP-Italic.otf'
        boldItalic: '/fonts/NotoSerifJP-BoldItalic.otf'
    }
  },
  {
    fallback: false,
    label: 'NotoSansJP',
    url: '/fonts/NotoSansJP-Regular.otf',
    // no variants avaiable for this font
  },
  {
    fallback: false,
    label: DEFAULT_FONT_NAME,
    data: getDefaultFont()[DEFAULT_FONT_NAME].data,
  },
];
```

TextSchema modifications
```ts
{
    // ... other properties,
    bold?: boolean, 
    // or
    fontWeight?: 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900,
    italic?: boolean
}
```

**Reply by hand-dot** [reply]
@AyushBobale 
Great suggestion! I think the approach of including font variants is really interesting. The interface also looks nice and simple.

As for related feature requests, we have the following issues:  
- https://github.com/pdfme/pdfme/issues/564
- https://github.com/pdfme/pdfme/issues/566

If we ultimately add a schema for rich-text fields, that would be truly amazing. However, it’s likely going to be a long-term project.

I’d like to explore the possibilities, but this could get lengthy. Could you create a new issue for this?  
It might be best to break the implementation down into several phases.

**Reply by hand-dot** [reply]
```
// variants would be optional
// if a variant is not found fallback to regular font
const fontObjList = [
  {
    fallback: true,
    label: 'NotoSerifJP',
    url: '/fonts/NotoSerifJP-Regular.otf'
    variants: {
        bold: '/fonts/NotoSerifJP-Bold.otf'
        italic: '/fonts/NotoSerifJP-Italic.otf'
        boldItalic: '/fonts/NotoSerifJP-BoldItalic.otf'
    }
  },
  {
    fallback: false,
    label: 'NotoSansJP',
    url: '/fonts/NotoSansJP-Regular.otf',
    // no variants avaiable for this font
  },
  {
    fallback: false,
    label: DEFAULT_FONT_NAME,
    data: getDefaultFont()[DEFAULT_FONT_NAME].data,
  },
];
```

ref: https://pdfmake.github.io/docs/0.1/fonts/custom-fonts-client-side/vfs/

**Reply by hand-dot** [reply]
Also, I think satori's font interface become an example.
https://github.com/vercel/satori/blob/main/README.md

**Comment by nicolaspaillard** [comment]
Hey @hand-dot, 
Implementing your awesome library, i needed a schema to add hyperlinks on text.
Its not using any external library, and uses the same prop panel and parameters as text.

## guide

create schemas directory where you need it and copy anything from `node_modules\@pdfme\schemas\dist\types\src`
create `schemas/link/link.ts`
![image](https://github.com/user-attachments/assets/e5348191-db72-41dd-892e-69219d64298e)
copy the following code in `link.ts`
```ts
import { PDFRenderProps, Plugin } from "@pdfme/common";
import { PDFString } from "@pdfme/pdf-lib";
import { Link } from "lucide";
import text from "../text";
import { pdfRender as parentPdfRender } from "../text/pdfRender";
import { TextSchema } from "../text/types";
import { convertForPdfLayoutProps, createSvgStr } from "../utils";

// inputs : [["text","url"]]

export interface LinkSchema extends TextSchema {
  url: string;
}

export const link: Plugin<LinkSchema> = {
  ui: text.ui,
  pdf: async (arg: PDFRenderProps<LinkSchema>) => {
    const { value, pdfDoc, schema, page, ...rest } = arg;
    const pageHeight = page.getHeight();
    const {
      width,
      height,
      position: { x, y },
    } = convertForPdfLayoutProps({ schema, pageHeight, applyRotateTranslate: false });
    const values: string[][] = JSON.parse(value);
    page.node.addAnnot(
      pdfDoc.context.register(
        pdfDoc.context.obj({
          Type: "Annot",
          Subtype: "Link",
          Rect: [x, y, x + width, y + height],
          // Border: [0, 0, 0],
          // border color
          // C: [0, 0, 1],
          // URI action
          A: {
            Type: "Action",
            S: "URI",
            URI: PDFString.of(values[0][1]),
            // still searching for that one
            // target: "_blank", 
          },
        }),
      ),
    );
    const renderArgs = {
      value: values[0][0],
      pdfDoc: pdfDoc,
      schema,
      page: page,
      ...rest,
    };

    await parentPdfRender(renderArgs);
  },
  propPanel: {
    schema: text.propPanel.schema,
    defaultSchema: {
      ...text.propPanel.defaultSchema,
      rotate: undefined,
      type: "link",
      url: "",
    },
  },
  icon: createSvgStr(Link),
};

export default link;

```
edit `schemas/index.ts` to make it export the new plugin
```ts
import barcodes from "./barcodes/index";
import checkbox from "./checkbox/index";
import date from "./date/date";
import dateTime from "./date/dateTime";
import time from "./date/time";
import image from "./graphics/image";
import svg from "./graphics/svg";
// link added
import link from "./link/link";
import multiVariableText from "./multiVariableText/index";
import radioGroup from "./radioGroup/index";
import select from "./select/index";
import line from "./shapes/line";
import { ellipse, rectangle } from "./shapes/rectAndEllipse";
import table from "./tables/index";
import text from "./text/index";

const builtInPlugins = { Text: text };

export {
  barcodes,
  builtInPlugins,
  checkbox,
  date,
  dateTime,
  ellipse,
  image,
  line,
  multiVariableText,
  radioGroup,
  rectangle,
  select,
  svg,
  table,
  text,
  time,
  // link added
  link,
};
```
hover on a generated link...
![image](https://github.com/user-attachments/assets/f850b99e-1327-4f56-8e1a-54a8396c7c80)

**Reply by hand-dot** [reply]
@nicolaspaillard 
Great work!

I’ve been considering adding a schema like this for quite some time:
https://github.com/pdfme/pdfme/issues/319

You’ve almost finished the job. Thanks!

One thing I’m curious about is the way links are being added.

What do you think about merging this link schema into the existing text schema and adding the link as one of its properties?

It wouldn’t allow links to be set via forms, but since users don’t typically add links themselves, as long as the designer can set it, isn’t that sufficient?

**Reply by nicolaspaillard** [reply]
> I’ve been considering adding a schema like this for quite some time:
https://github.com/pdfme/pdfme/issues/319

Ahah i used the exact same comment from hopding 

> One thing I’m curious about is the way links are being added.

The way links are added is it adds an annotations layer over the whole pdf and puts links there, that's why i had to use addAnnot instead of set because set updates all annotations at once, so it was always removing previous ones.

> What do you think about merging this link schema into the existing text schema and adding the link as one of its properties?

Greeat idea ! I made it totally compatible by default but didnt think of it like this. The rotation may have to be disabled if a link is set tho, because i don't know how to rotate annotation. 

> It wouldn’t allow links to be set via forms, but since users don’t typically add links themselves, as long as the designer can set it, isn’t that sufficient?

I may try to code this next week, i'm also thinking about adding kind of a popup input (appearing when the field is focused) to fill the link url in the form

Meanwhile, anyone can already implement or modify this 
To me the form inputs are useless but it may come handy in a future project i'll make

**Reply by nicolaspaillard** [reply]
Just want to add that annotations in pdf-lib are not documented and reversing it is a nightmare

**Comment by SHUHAIB-T** [comment]
# Feature Request: Enhanced Text Styling Options for PDFme

Currently, PDFme's text schema lacks built-in support for rich text formatting, which limits design flexibility when generating PDF templates. I propose adding comprehensive text styling options that would include:

## 1. Text Formatting Capabilities
- Bold text styling
- Italic text styling
- Underline text styling
- Combination of formatting (e.g., bold italic)

## 2. Potential Implementation Approaches
```javascript
// Example proposed schema structure
{
  type: 'text',
  text: 'Sample Text',
  fontStyle: {
    bold: true,
    italic: false,
    underline: false
  }
}
```

## 3. Key Benefits
- More dynamic text rendering in PDF templates
- Increased design flexibility
- Simplified text styling workflow
- Consistent rendering across UI and PDF generation

## Technical Considerations
- Extend existing text schema plugin
- Modify UI and PDF rendering methods
- Ensure compatibility with current PDFme architecture
- Provide backward compatibility with existing templates

## Example Implementation Prototype
```javascript
const richTextPlugin = {
  ui: (props) => {
    const { schema, value } = props;
    const span = document.createElement('span');
    span.textContent = value;
    
    // Apply styling based on fontStyle configuration
    span.style.fontWeight = schema.fontStyle?.bold ? 'bold' : 'normal';
    span.style.fontStyle = schema.fontStyle?.italic ? 'italic' : 'normal';
    span.style.textDecoration = schema.fontStyle?.underline ? 'underline' : 'none';
  },
  pdf: (props) => {
    // Custom PDF rendering logic
  }
};
```

## Questions for Community
- Would this approach meet your text styling requirements?
- What additional formatting options would be valuable?
- Are there specific use cases I might have overlooked?

*I'm happy to collaborate on refining this proposal and potentially contribute a pull request if there's community interest.*

**Reply by hand-dot** [reply]
Hey @SHUHAIB-T 
related this: https://github.com/pdfme/pdfme/discussions/288#discussioncomment-11528928

**Comment by Myshanyatko** [comment]
### Custom plugin for QR-code (using pdfmake)

https://gist.github.com/Myshanyatko/f0fca460ae8f1bb6aa3e250362b623c8

This plugin was created because I needed QR code rendering in pdfme, but didn't want to depend on libraries like antd, form-render, and react that are used by @pdfme/schemas for QR codes. This plugin uses the pdfmake library to render QR codes, and it embeds them into your PDF as PDF-formatted content. This solution was convenient for me because the pdfmake library was already installed in the project. :)
The plugin is intended only for PDF generation — no UI features are included!

'+': Draw qr-code without using new heavy versions of `@pdfme/shemas`
'-': There is a dependency on the `pdfmake` library

**Comment by tetty0217** [comment]
# Proposed countermeasure for overflow generated by TextSchema

Hi @hand-dot

Today I would like to propose a feature in pdfme text display that does not display strings that are out of bounds.

## Summary

**Problem**

Text in `TextSchema` is currently allowed to extend beyond the container, but this does not guarantee that the output will follow a predefined layout.

We feel that the current situation is a challenge for documents where layout is important, such as invoices and estimates, where text processing and validation cannot follow the pdfme rendering logic.

**Proposal**

 I would like to add a new boolean option to `TextSchema`. This option would explicitly control text overflow and ensure that the rendered output always respects the defined bounds.

**advantage**

- For developers
  - Provides developers with basic control over text rendering and can guarantee layout consistency. 
- For the end user
  - Can ensure that generated documents accurately reflect user input in the intended design.

<img src="https://github.com/user-attachments/assets/8206f1b7-1708-453b-a118-5237aa8ad117" width="400" alt="before"/>
<img src="https://github.com/user-attachments/assets/91a46e83-9415-47e8-a493-e451e4b5cf2c" width="400" alt="after"/>


## Detail
### 1. Problem to be solved (Problem / Motivation)

#### Current Problem Points:

The TextSchema overflows the string from the frame, but there doesn't seem to be a way to prevent this in the pdfme.

I would like to Contribute, but there may be an idea that it should be a custom schema or Fork depending on the content, so I will post it in Discussion.

#### Specific Use Case:

I have set DynamicFontSize in TextSchema to allow users to type more characters (this is a very nice feature!!!).

However, I am having usability and technical challenges when creating templates for official documents such as invoices, quotations, and contracts.

**1. issue from the user's point of view**

Over-extending the size of TextSchema to accommodate more text is not very practical for formal documents such as ledgers.

Therefore, an approach was needed that would allow DynamicFontSize to work, but still keep the text completely within the boundaries of the box.


**2. Technical issues**

It is “technically possible” to truncate or validate the string input to Template based on the number of characters or lines so that the input string fits within the frame.

However, since pdfme internally (pdfRender) manages the strings to be displayed as “lines,” duplicating the logic involved in PDF rendering resulted in poor performance with Rutime.

Furthermore, many other helpers and utilities are not exported from pdfme, so it is currently difficult to process and validate text using the same display decisions as pdfme by means other than Fork.

---

### 2. Proposed Solution

To solve this problem, we propose to “add an option to TextSchema to not display overflow strings”.

#### Specifically
- Add a flag (e.g. respectTextBounds) to TextSchema to not display strings according to the height of TextSchema.
- In uiRender, if this flag is enabled, set `overflow: ‘hidden’` in the TextSchema container to not display strings that extend beyond the bounds.
- To avoid displaying truncated out-of-bounds text when this flag is enabled, we believe pdfRender would use size calculations to truncate out-of-bounds “line” candidates and then draw the borders and text via the pdfLib feature.

#### Advantages of this solution
- For developers
  - Provides developers with basic control over text rendering and can guarantee layout consistency.
- For end users
  - Ensures that the generated documents accurately reflect user input in the intended design.

### 3. Alternatives considered

In addition to this proposal, the following alternatives were also considered

#### **Alternative:** Implement a TextSchema wrapper for product use cases.

**Reason not selected:**

As mentioned above, this alternative was not selected because the logic of interest, PDF rendering for Input, is covered by the internal implementation.

It is possible to approximate text parsing and processing with Intl.Segmenter and pdfme/common so that the input string can be displayed as is, but the performance degradation and DynamicFontSize support made this difficult. 

Therefore, we believe our proposal is the most balanced at this time.

--- 

I'd appreciate your feedback on this proposal. 

If you agree with this direction, I'm happy to work on a pull request.

**Reply by hand-dot** [reply]
Hi @tetty0217 
Why wouldn't setting the DynamicFontSize Min to something like 0.1 work? I think it's probably not a great option since the text would be unreadable in that case, but it would function in terms of preventing layout issues.

**Comment by cagan123-alt** [comment]
I found in general using tables but sometimes you need some custom logic.

Example:
You have customers you want to print in a list but this list contains a lot of customers and depending on the template you want to render the ones with payment more than 1000 dollars.

What i found is custom table that will give user

schema spesficiation

Example

User selects Custom table

in code user can defines their arrays
like let's assume user has 3 arrays that will be used they are

Customers
Products
Vacations


user defines them
Customer -> contains name surname phone and responsiblePerson
Vacation -> startDate endDate createdAt


this table we have will only has 1 row user will see and can't have more

and in this part user can add headers and reference values from the given schema

Example header for customer

Customer Name
Customer Surname


then user can go ahead and in the 1 row they have go

${customer.name}
${customer.surname}

// Optinally null and undefined behavior also can be added

and then goes ahead and writes json-rules-engine code directly given input field in the custom rules section in the table

with this custom rules we can spesify only render customer that starts with C or T etc...

With this assume that you have a transaction list that has 35 columns in your database and want to make so users can customize what is printed

You will define it and all its columns

User goes ahead and defines what it want maybe they just wants to show amount and currency and nothing more

they can just add them

add their own custom rules
like only show transactions that's bigger than 1000
then it's done user can configure what they want to render and the api doesn't need to change for anything user can do and render whatever they want however they want and we can configure it depending on our data

Maybe its vacations we can create it depending on our vacation data in code like i explained

**Reply by cagan123-alt** [reply]
I think this would be really cool to do but not in the current plugin system but in a shadcn approach where it gives you the code of the plugin with inital schemas and then you configure it depending on your need

**Comment by MarioWork** [comment]
## Feature Request: Page Options for PDFme UI
 With the current version of **Pdfme UI**, it is **not possible** to add a new page with:
 
 * A specific size (e.g., A4, A5, etc.)
 * A specific orientation (landscape or portrait)
 
 ## Description of the feature
 It would be awesome to add the ability to specify page options when adding a new page, such as:
 
 * Page size (e.g., A4, A5, Letter, etc.)
 * Page orientation (portrait or landscape)
 
 This enhancement would provide greater flexibility and control over PDF creation within the UI.
 
 ## Possible Solutions
 * **Add New Page Popup:**
   When the user clicks "Add New Page," open a popup dialog styled like PDFme UI that allows selecting the page size and orientation before creating the page.
 * **Default Page with Editable Options:**
   Create the new page with a default size and orientation, then provide a section in the fields sidebar where users can adjust the page options (size and orientation) after the page is added.

**Reply by MarioWork** [reply]
#1156

**Comment by moelfus** [comment]
### Embed PDF Image
For the automatic creation of data sheets, I need to embed CAD drawings. These drawings are provided as PDFs, and I initially tried converting them to SVG using muPDF and then embedding them via the SVG plugin. Unfortunately, that approach did not work reliably: in some cases only the text was visible, in others only the drawing, and sometimes both appeared but were rendered on top of each other incorrectly. 
As a result, I explored the image scheme and the QR plugin and implemented a custom scheme that embeds PDF drawings directly—essentially treating them the same way as images.
I’m not a skilled programmer, but I wanted to share this code as a starting point for others who might face a similar challenge. You can find the code [here](https://gist.github.com/moelfus/35eb54cd01f7d3d30e08683fc6631c03).

For testing:
- just place the plugin somewhere in your project and reference it via `import pdfImageSchema from './PDFImage'`; (adjust path)
- for the UI or generator add it to the plugins section in you code: const plugins = { text, image, qrcode: barcodes.qrcode, svg, table, PDFImage: pdfImageSchema };
